### PR TITLE
Adding IdentifiableSecrets helper

### PR DESCRIPTION
**Pull request by eddynaka**
# Description

This change will introduce two new public APIs:
- GenerateIdentifiableKey
- ValidateKey

### CODE REVIEW

File: src/Directory.Build.props [hunk 1]
```diff
@@ -2,6 +2,7 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
 
   <PropertyGroup Label="Build">
+    <LangVersion>8</LangVersion>
```
Comment (eddynaka) [inline src/Directory.Build.props:5]: ><LangVersion>8</LangVersion> [](http://example.com/codeflow?start=4&length=28)

this is required to use the new `using` format.

File: src/Tests.Marvin/IdentifiableSecretsTests.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+{
+    [TestClass]
+    public class IdentifiableSecretsTests
+    {
+        private const string Signature = "+Sig";
+
+        [TestMethod]
+        public void IdentifiableSecrets_GenerateIdentifiableKey_ShouldThrowExceptionWhenLengthIsInvalid()
+        {
+            ulong seed = BitConverter.ToUInt64(Encoding.ASCII.GetBytes("default-seed").Reverse().ToArray(), 0);
```
Comment (michaelcfanning) [inline src/Tests.Marvin/IdentifiableSecretsTests.cs:16]: >default-seed [](http://example.com/codeflow?start=72&length=12)

Use a constant literal for this. Please note that only 8 characters are interpreted for the seed, so it's confusing that you provided a 12-byte value. Use "DEFAULT0", perhaps.

File: src/Tests.Marvin/IdentifiableSecretsTests.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            var isValid = IdentifiableSecrets.ValidateKey(newSecret, seed, Signature);
+            Assert.IsFalse(isValid);
+        }
+
+        [TestMethod]
+        public void IdentifiableSecrets_GenerateIdentifiableKey_ShouldBeValidKey()
+        {
+            ulong seed = BitConverter.ToUInt64(Encoding.ASCII.GetBytes("default-seed").Reverse().ToArray(), 0);
+
+            var sizes = new uint[] { 24, 28, 32, 36, 40 };
```
Comment (michaelcfanning) [inline src/Tests.Marvin/IdentifiableSecretsTests.cs:50]: >24 [](http://example.com/codeflow?start=37&length=2)

These sizes aren't super interesting, as they are all divisible by four, so we won't cover all corner cases. Did you verify you're achieving 100% code coverage?

I'd suggest testing the minimal key size, the maximal key size and at least 4 contiguous sizes, e.g., 63, 64, 65, and 66.

File: src/Tests.Marvin/IdentifiableSecretsTests.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                IdentifiableSecrets.GenerateIdentifiableKey(seed, keyLengthInBytes: 32, base64EncodedSignature: null));
+
+            Assert.ThrowsException<ArgumentException>(() =>
+                IdentifiableSecrets.GenerateIdentifiableKey(seed, keyLengthInBytes: 32, base64EncodedSignature: "bad-signature-length"));
+        }
+
+        [TestMethod]
+        public void IdentifiableSecrets_ValidateKey_ShouldReturnFalseIfSecretIsInvalid()
+        {
+            const string reversedSignature = "giS+";
```
Comment (michaelcfanning) [inline src/Tests.Marvin/IdentifiableSecretsTests.cs:34]: >"giS+" [](http://example.com/codeflow?start=45&length=6)

Why don't you actually reverse the sig?

Encoding.ASCII.GetBytes(Signature).Reverse()

You'll need to convert back to string of course.

Honestly, any signature that differs from Signature will be ok.

File: src/Tests.Marvin/IdentifiableSecretsTests.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+            Assert.ThrowsException<ArgumentException>(() =>
+                IdentifiableSecrets.GenerateIdentifiableKey(seed, keyLengthInBytes: 32, base64EncodedSignature: null));
+
+            Assert.ThrowsException<ArgumentException>(() =>
+                IdentifiableSecrets.GenerateIdentifiableKey(seed, keyLengthInBytes: 32, base64EncodedSignature: "bad-signature-length"));
+        }
+
+        [TestMethod]
+        public void IdentifiableSecrets_ValidateKey_ShouldReturnFalseIfSecretIsInvalid()
```
Comment (michaelcfanning) [inline src/Tests.Marvin/IdentifiableSecretsTests.cs:32]: >IdentifiableSecrets_ValidateKey_ShouldReturnFalseIfSecretIsInvalid [](http://example.com/codeflow?start=20&length=66)

You didn't test passing the wrong checksum.

File: src/Marvin/IdentifiableSecrets.cs [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+                        // bits, the first encoded character of the checksum will have its
+                        // first two bits set to zero, limiting encoded chars to A-P.
+                        if (firstChar < 'A' || firstChar > 'P')
+                        {
+                            return false;
+                        }
+
+                        pattern = $"[0-9a-zA-Z\\/\\+]{{{prefixLength}}}{expectedSignature}[A-P][0-9a-zA-Z\\/\\+]{{5}}{equalsSigns}";
+                        break;
+                    }
```
Comment (michaelcfanning) [inline src/Marvin/IdentifiableSecrets.cs:134]: these blocks shouldn't be indented, you need to change your VS editor defaults for this.

File: src/Marvin/IdentifiableSecrets.cs [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+            }
+
+            var regex = new Regex(pattern);
+            return regex.IsMatch(key);
+        }
+
+        private static string GenerateKeyWithAppendedSignatureAndChecksum(
+            byte[] keyValue,
```
Comment (michaelcfanning) [inline src/Marvin/IdentifiableSecrets.cs:172]: >yte[] keyValue, [](http://example.com/codeflow?start=13&length=15)

this isn't our parameter formatting standard, please add the first arg to the previous line, then line up all params beneath it.

File: src/Marvin/IdentifiableSecrets.cs [hunk 8]
```diff
@@ -0,0 +1,270 @@
+﻿using System;
+using System.Security.Cryptography;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Security.Utilities
+{
+    public static class IdentifiableSecrets
+    {
+        private static uint MaximumGeneratedKeySize => 4096;
```
Comment (michaelcfanning) [inline src/Marvin/IdentifiableSecrets.cs:9]: >private [](http://example.com/codeflow?start=8&length=7)

Make these internal so tests can use them for boundary conditions.

**Review by michaelcfanning — CHANGES_REQUESTED** [review]
:clock1:

File: src/Marvin/IdentifiableSecrets.cs (hunk 6 above)
Comment (eddynaka) [reply src/Marvin/IdentifiableSecrets.cs:134]: fixed.

File: src/Marvin/IdentifiableSecrets.cs (hunk 7 above)
Comment (eddynaka) [reply src/Marvin/IdentifiableSecrets.cs:172]: changed.

File: src/Tests.Marvin/IdentifiableSecretsTests.cs (hunk 5 above)
Comment (eddynaka) [reply src/Tests.Marvin/IdentifiableSecretsTests.cs:41]: just added one :)

File: src/Tests.Marvin/IdentifiableSecretsTests.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            const int size = 32;
+            ulong seed = BitConverter.ToUInt64(Encoding.ASCII.GetBytes(Seed).Reverse().ToArray(), 0);
+
+            string secret = IdentifiableSecrets.GenerateIdentifiableKey(seed,
+                                                                        keyLengthInBytes: size,
+                                                                        base64EncodedSignature: Signature);
+
+            string newSecret = secret.Replace(Signature, Signature.ToLower());
```
Comment (michaelcfanning) [inline src/Tests.Marvin/IdentifiableSecretsTests.cs:50]: >Signature [](http://example.com/codeflow?start=57&length=9)

If you take this approach you need to add a new assert that verifies our Signature != Signature.ToLowerInvariant().

Make sense? Then the test will fail if someone changes Signature to be an all lower-case value.

Also, please don't use ToLower() in general, it is a culture-sensitive helper, ToLowerInvariant() would be preferred.

**Review by michaelcfanning — APPROVED** [review]
:shipit:
